### PR TITLE
DOCSUP-5920: Documented PostgreSQL engines, function and dictionary

### PR DESCRIPTION
**Pull request by olgarev**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- Documentation for #18554, #21710, #21711

### CODE REVIEW

**Comment by kssenii** [comment]
FYI there is another PR with PostgreSQL doc https://github.com/ClickHouse/ClickHouse/pull/21220. I believe it should be merged first. May be after you will just correct and combine the two docs together? (There is no doc for database engine in that pr, which is added here, and other corrections and usage examples)

**Comment by olgarev** [comment]
> FYI there is another PR with PostgreSQL doc #21220. I believe it should be merged first. May be after you will just correct and combine the two docs together? (There is no doc for database engine in that pr, which is added here, and other corrections and usage examples). May be you can help figure out what is wrong with `docs check` in that pr?

@kssenii We have such recommendation in out Troubleshooting wiki:

image

If you add a new file, it should end with a string like:

[Original article](https://clickhouse.tech/docs/<path-to-the-page>) 

and it should be a new empty line after it.

May be this will help.

File: docs/en/sql-reference/table-functions/index.md [hunk 1]
```diff
@@ -21,17 +21,18 @@ You can use table functions in:
...
+| [jdbc](../../sql-reference/table-functions/jdbc.md)              | Creates a [JDBC](../../engines/table-engines/integrations/jdbc.md)-engine table.                                                       |
+| [odbc](../../sql-reference/table-functions/odbc.md)              | Creates a [ODBC](../../engines/table-engines/integrations/odbc.md)-engine table.                                                       |
+| [hdfs](../../sql-reference/table-functions/hdfs.md)              | Creates a [HDFS](../../engines/table-engines/integrations/hdfs.md)-engine table.                                                       |
+| [s3](../../sql-reference/table-functions/s3.md)                  | Creates a [S3](../../engines/table-engines/integrations/s3.md)-engine table.                                                           |
+| [postgresql](../../sql-reference/table-functions/postgresql.md)  | Creates a [PostgreSQL](../../engines/table-engines/integrations/postgresql.md)-engine table.                                           |
 
 [Original article](https://clickhouse.tech/docs/en/query_language/table_functions/) <!--hide-->
```
Comment (adevyatova) [inline docs/en/sql-reference/table-functions/index.md:38]:
```suggestion
[Original article](https://clickhouse.tech/docs/en/sql-reference/table-functions/) <!--hide-->
```

File: docs/en/sql-reference/table-functions/postgresql.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+│ 7 │ 8 │ 9 │
+└───┴───┴───┘
+```
+
+**See Also**
+
+-   [The PostgreSQL table engine](../../engines/table-engines/integrations/postgresql.md)
+-   [Using PostgreSQL as a source of external dictionary](../../sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md#dicts-external_dicts_dict_sources-postgresql)
+
+[Original article](https://clickhouse.tech/docs/en/sql-reference/table_functions/postgresql/) <!--hide-->
```
Comment (adevyatova) [inline docs/en/sql-reference/table-functions/postgresql.md:73]:
```suggestion
[Original article](https://clickhouse.tech/docs/en/sql-reference/table-functions/postgresql/) <!--hide-->
```

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md [hunk 3]
```diff
@@ -673,4 +675,52 @@ Default value is 1 (the first key column is a partition key and other key column
...
+
+- `db` — Remote database name.
+- `host` — The PostgreSQL host.
+- `port` – The port on the PostgreSQL server. If not specified, default port is used.
+- `user` — PostgreSQL user.
+- `password` — User password.
+- `table` — Remote table name.
+- `invalidate_query` — Query for checking the dictionary status. Optional parameter. Read more in the section [Updating dictionaries](../../../sql-reference/dictionaries/external-dictionaries/external-dicts-dict-lifetime.md).
+
 [Original article](https://clickhouse.tech/docs/en/query_language/dicts/external_dicts_dict_sources/) <!--hide-->
```
Comment (adevyatova) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md:726]:
```suggestion
[Original article](https://clickhouse.tech/docs/en/sql-reference/dictionaries/external-dictionaries/external-dict-sources) <!--hide-->
```

File: docs/en/engines/database-engines/postgresql.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+DESCRIBE TABLE test_database.test_table;
+```
+``` text
+┌─name───┬─type──────────────┐
+│ id     │ Nullable(Integer) │
+│ value  │ Nullable(Integer) │
+└────────┴───────────────────┘
+```
+
+After detaching the table and attaching it again, the structure was updated:
```
Comment (adevyatova) [inline docs/en/engines/database-engines/postgresql.md:123]:
```suggestion
After detaching the table and attaching it again, the structure will be updated:
```

**Comment by olgarev** [comment]
@kssenii Hi! Maybe it's better to merge my PR first, and then merge the changes from the second PR into it?

**Comment by kssenii** [comment]
> @kssenii Hi! Maybe it's better to merge my PR first, and then merge the changes from the second PR into it?

Hi! It seems too complicated to make any more changes to that second PR. Lets wait some more time, please.

**Comment by kssenii** [comment]
@olgarev, PR is fixed and merged. Can you, please, merge the changes?

Also, there are two PR's which also need PostgreSQL doc :)
https://github.com/ClickHouse/ClickHouse/pull/21711 (additional parameter `table_schema`)
https://github.com/ClickHouse/ClickHouse/pull/21710 (same as now supported for mysql)

**Comment by olgarev** [comment]
> @olgarev, PR is fixed and merged. Can you, please, merge the changes?

@kssenii Great! Will do my best )

> Also, there are two PR's which also need PostgreSQL doc :)
> #21711 (additional parameter `table_schema`)
> #21710 (same as now supported for mysql)

If you tag them with "docs-alert", we will pick them when ready. Thank you!

**Comment by olgarev** [comment]
@kssenii Hi, wil you review this PR, pls?

File: docs/ru/engines/database-engines/postgresql.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+toc_title: PostgreSQL
+---
+
+# PostgreSQL {#postgresql}
+
+Позволяет подключаться к БД на удаленном сервере [PostgreSQL](https://www.postgresql.org). Поддерживает операции чтения и записи (запросы `SELECT` и `INSERT`) для обмена данными между ClickHouse и PostgreSQL.
+
+Позволяет в реальном времени получать от удаленного сервера PostgreSQL информацию о таблицах БД и их структуре с помощью запросов `SHOW TABLES` и `DESCRIBE TABLE`.
+
+Поддерживает операции изменения структуры таблиц (`ALTER TABLE ... ADD|DROP COLUMN`). Если параметр `use_table_cache` (см. ниже раздел Параметры движка) установлен в значение `1`, структура таблицы кешируется, и изменения в структуре не отображаются мгновенно, но могут быть получены с помощью запросов `DETACH` и `ATTACH`.
```
Comment (kssenii) [inline docs/ru/engines/database-engines/postgresql.md:12]:
```suggestion
Поддерживает операции изменения структуры таблиц (`ALTER TABLE ... ADD|DROP COLUMN`). Если параметр `use_table_cache` (см. ниже раздел Параметры движка) установлен в значение `1`, структура таблицы кешируется, и изменения в структуре не отслеживаются, но будут обновлены, если выполнить команды `DETACH` и `ATTACH`.
```

File: docs/en/engines/table-engines/integrations/postgresql.md [hunk 6]
```diff
@@ -29,25 +29,28 @@ The table structure can differ from the original PostgreSQL table structure:
...
+`INSERT` queries on PostgreSQL side run as `COPY "table_name" (field1, field2, ... fieldN) FROM STDIN` inside PostgreSQL transaction with auto-commit after each `INSERT` statement.
+
+PostgreSQL `Array` types are converted into ClickHouse arrays.
+
+!!! info "Note"
+    Be careful - in PostgreSQL an array data, created like a `type_name[]`, may contain multi-dimensional arrays of different dimensions in different table rows in same column. But in ClickHouse it is only allowed to have multidimensional arrays of the same count of dimensions in all table rows in same column.
 
-PostgreSQL Array types converts into ClickHouse arrays.
-Be careful in PostgreSQL an array data created like a type_name[] may contain multi-dimensional arrays of different dimensions in different table rows in same column, but in ClickHouse it is only allowed to have multidimensional arrays of the same count of dimensions in all table rows in same column.
+Supports replicas priority for PostgreSQL dictionary source. The bigger the number in map, the less the priority. The highest priority is `0`.
```
Comment (kssenii) [inline docs/en/engines/table-engines/integrations/postgresql.md:53]: An example of configuration with replicas can be added. 
  ```<source>
  <postgresql>
      <port>5432</port>
      <user>clickhouse</user>
      <password>qwerty</password>
      <replica>
          <host>example01-1</host>
          <priority>1</priority>
      </replica>
      <replica>
          <host>example01-2</host>
          <priority>2</priority>
      </replica>
      <db>db_name</db>
      <table>table_name</table>
      <where>id=10</where>
      <invalidate_query>SQL_QUERY</invalidate_query>
  </postgresql>
</source>```

Comment (olgarev) [reply docs/en/engines/table-engines/integrations/postgresql.md]: @kssenii Done

**Review by kssenii — APPROVED** [review]

**Pull request merged**